File: 142.detectCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == NULL) return NULL;

        ListNode* slow = head;
        ListNode* fast = head;

        int flag = -1;
        while(fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
            fast = fast->next;
            if (fast == slow) {
                flag = 1;
                break;
            }
        }

        if (flag == -1) return NULL;

        fast = slow;
        slow = head;
        while(fast != slow) {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};

File: 142.detectCycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func detectCycle(head *ListNode) *ListNode {
    first := head
    second := head

    for first != nil{
        second = second.Next
        if first.Next == nil {
            return nil
        }
        first = first.Next.Next

        if first == second {
            first = head
            for first != second {
                second = second.Next
                first = first.Next
            }
            return first
        }
    }

    return nil
}